content/6_Advanced/MinCostFlow.mdx

```diff
@@ -211,19 +211,105 @@ and
+$i$th job to the $j$th worker is $C_{i,j}$.
 
-Better resources? I honestly just found some that came up in a google search.
+We will create a flow network with a source node $S$, job nodes $J_i$, worker
+nodes $W_j$, and a sink node $E$.
 
-</IncompleteSection>
+We will have the following edges:
+
+- $(S, J_i)$ with $(capacity, cost) = (1, 0)$
+- $(J_i, W_j)$ with $(1, C_{i,j})$
+- $(W_j, E)$ with $(1, 0)$
+
+The answer will be the minimum cost of the max flow of the graph. An example:
+
+![Hungarian Algorithm Flows Diagram](./hungarian_algo_graph.png)
+
+To solve this, we will use the min cost max flow algorithm detailed above. In every iteration
+of the algorithm, we will assign a job to a worker. In each iteration, we will
+first assign the job to an auxilary worker, then we will try to find an
+augmenting path with minimum cost. To optimize this, we will only run Dijkstra
+on the worker nodes. For details, refer to the code.
+
+Since there are $J$ iterations (for each job) and each iteration takes $O(W^2)$
+time, the overall time complexity is $O(J W^2)$.
+
+### Implementation
+
+From Wikipedia (which is then copied from e-maxx):
+
+```cpp
+/**
+ * @return the jobs of each worker in the optimal assignment,
+ * or -1 if the worker is not assigned
+ */
+template <class T> vector<T> hungarian(const vector<vector<T>> &C) {
+	int J = C.size();
+	int W = C[0].size();
+	assert(J <= W);
+
+	// job[w] = job assigned to w-th worker, or -1 if no job assigned
+	// note: a W-th worker was added for convenience
+	vector<int> job(W + 1, -1);
+	vector<T> h(W);  // Johnson potentials
+
+	const T inf = numeric_limits<T>::max();
+	// assign j_cur-th job using Dijkstra with potentials
+	for (int j_cur = 0; j_cur < J; j_cur++) {
+		int w_cur = W;  // unvisited worker with minimum distance
+		job[w_cur] = j_cur;
+
+		vector<T> dist(W + 1, inf);  // Johnson-reduced distances
+		dist[W] = 0;
+		vector<bool> vis(W + 1);     // whether visited yet
+		vector<int> prv(W + 1, -1);  // previous worker on shortest path
+		while (job[w_cur] != -1) {   // Dijkstra step: pop min worker from heap
+			T min_dist = inf;
+			vis[w_cur] = true;
+			int w_next = -1;  // next unvisited worker with minimum distance
+
+			// consider extending shortest path by w_cur -> job[w_cur] -> w
+			for (int w = 0; w < W; w++) {
+				if (!vis[w]) {
+					// sum of reduced edge weights w_cur -> job[w_cur] -> w
+					T edge = C[job[w_cur]][w] - h[w];
+					if (w_cur != W) {
+						edge -= C[job[w_cur]][w_cur] - h[w_cur];
+						assert(edge >= 0);
+					}
+					if (ckmin(dist[w], dist[w_cur] + edge)) { prv[w] = w_cur; }
+					if (ckmin(min_dist, dist[w])) { w_next = w; }
+				}
+			}
+			w_cur = w_next;
+		}
+
+		for (int w = 0; w < W; w++) {  // update potentials
+			ckmin(dist[w], dist[w_cur]);
+			h[w] += dist[w];
+		}
+
+		while (w_cur != W) {  // update job assignment
+			job[w_cur] = job[prv[w_cur]];
+			w_cur = prv[w_cur];
+		}
+	}
+
+	return job;
+}
```

Comment:
maybe put a `main` function with sample input to `C` to people can see this method in action